test/boost/lister_test.cc

```diff
@@ -0,0 +1,183 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <unordered_set>
+#include <fmt/format.h>
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/file.hh>
+
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include "test/lib/tmpdir.hh"
+#include "test/lib/random_utils.hh"
+#include "test/lib/make_random_string.hh"
+
+#include "lister.hh"
+
+class expected_exception : public std::exception {
+public:
+    virtual const char* what() const noexcept override {
+        return "expected_exception";
+    }
+};
+
+static future<> touch_file(fs::path filename, open_flags flags = open_flags::wo | open_flags::create, file_permissions create_permissions = file_permissions::default_file_permissions) {
+    file_open_options opts;
+    opts.create_permissions = create_permissions;
+    auto f = co_await open_file_dma(filename.native(), flags, opts);
+    co_await f.close();
+}
+
+SEASTAR_TEST_CASE(test_empty_queue_lister) {
+    auto tmp = tmpdir();
+    auto ql = queue_lister(tmp.path());
+    size_t count = 0;
+
+    while (auto de = co_await ql.get()) {
+        count++;
+    }
+
+    BOOST_REQUIRE(!count);
+}
+
+static future<size_t> generate_random_content(tmpdir& tmp, std::unordered_set<std::string>& file_names, std::unordered_set<std::string>& dir_names, size_t max_count = 1000) {
+    size_t count = tests::random::get_int<size_t>(0, max_count);
+    for (size_t i = 0; i < count; i++) {
+        auto name = tests::random::get_sstring(tests::random::get_int(1, 8));
+        if (tests::random::get_bool()) {
+            if (!file_names.contains(name)) {
+                dir_names.insert(name);
+            }
+        } else {
+            if (!dir_names.contains(name)) {
+                file_names.insert(name);
+            }
+        }
+    }
+
+    for (const auto& name : file_names) {
+        co_await touch_file(tmp.path() / name);
+    }
+    for (const auto& name : dir_names) {
+        co_await touch_directory((tmp.path() / name).native());
+    }
+
+    co_return file_names.size() + dir_names.size();
+}
+
+SEASTAR_TEST_CASE(test_queue_lister) {
+    auto tmp = tmpdir();
+
+    std::unordered_set<std::string> file_names;
+    std::unordered_set<std::string> dir_names;
+
+    auto count = co_await generate_random_content(tmp, file_names, dir_names);
+    BOOST_TEST_MESSAGE(fmt::format("Generated {} dir entries", count));
+
+    std::unordered_set<std::string> found_file_names;
+    std::unordered_set<std::string> found_dir_names;
+
+    auto ql = queue_lister(tmp.path());
+
+    while (auto de = co_await ql.get()) {
+        assert(de->type);
```

Comment:
I don't think we should be using assert() in boost-tests. I think Boost has its own idioms for checks. Of course if the test doesn't fail, it doesn't matter... But if it does fail, we'll get a core dump instead of a normal boost test report, right?

Reply:
I didn't use `BOOST_REQUIRE` so not to spam the log with hundreds of boost messages.
If boost had a check that logs nothing by default I'd use it happily.

Reply:
using assert in tests works great in pytest, but in Boost tests, if this test ever fails, it will cause a crash instead of normal error messages. That just doesn't sound right.
I don't know about the spamming. Why should Boost log ever **successful** check? I don't remember that it does. Please re-check, and perhaps consult with people who are bigger fans than me with that test framework (I am *not* a fan).

Reply:
with `-l all` it prints successes as well. It's not the default though, but I use it in my scripts.

Reply:
So you use a non-default option to make tests more spammy, and then surprised when it's too spammy? :-)

Reply:
Fixed in v3